src/centrality.rs

```diff
@@ -0,0 +1,298 @@
+            do_scale = false;
+        } else {
+            scale = 1.0 / ((node_count - 1) * (node_count - 2)) as f64;
+        }
+    } else if !directed {
+        scale = 0.5;
+    } else {
+        do_scale = false;
+    }
+    if do_scale {
+        for x in betweenness.iter_mut() {
+            *x = x.map(|y| y * scale);
+        }
+    }
+}
+
+fn _accumulate_basic(
+    betweenness: &mut Vec<Option<f64>>,
+    max_index: usize,
+    path_calc: &mut ShortestPathData,
+    is: usize,
+) {
+    let mut delta = vec![0.0; max_index];
+    for w in &path_calc.verts_sorted_by_distance {
+        let iw = w.index();
+        let coeff = (1.0 + delta[iw]) / (path_calc.sigma[&iw] as f64);
+        let p_w = path_calc.predecessors.get(w).unwrap();
+        for v in p_w {
+            let iv = (*v).index();
+            delta[iv] += (path_calc.sigma[&iv] as f64) * coeff;
+        }
+        if iw != is {
+            betweenness[iw] = betweenness[iw].map(|x| x + delta[iw]);
+        }
+    }
+}
+
+fn _accumulate_endpoints(
+    betweenness: &mut Vec<Option<f64>>,
+    max_index: usize,
+    path_calc: &mut ShortestPathData,
+    is: usize,
+) {
+    betweenness[is] = betweenness[is]
+        .map(|x| x + ((path_calc.verts_sorted_by_distance.len() - 1) as f64));
+    let mut delta = vec![0.0; max_index];
+    for w in &path_calc.verts_sorted_by_distance {
+        let iw = w.index();
+        let coeff = (1.0 + delta[iw]) / (path_calc.sigma[&iw] as f64);
+        let p_w = path_calc.predecessors.get(w).unwrap();
+        for v in p_w {
+            let iv = (*v).index();
+            delta[iv] += (path_calc.sigma[&iv] as f64) * coeff;
+        }
+        if iw != is {
+            betweenness[iw] = betweenness[iw].map(|x| x + delta[iw] + 1.0);
+        }
+    }
+}
+
+struct ShortestPathData {
+    verts_sorted_by_distance: Vec<NodeIndex>,
+    predecessors: HashMap<NodeIndex, Vec<NodeIndex>>,
+    sigma: HashMap<usize, i64>,
+}
+
+fn shortest_path_for_centrality<G>(
+    graph: G,
+    node_s: &G::NodeId,
+) -> ShortestPathData
+where
+    G: NodeIndexable
+        + IntoNodeIdentifiers
+        + IntoNeighborsDirected
+        + NodeCount
+        + GraphBase<NodeId = NodeIndex>, // for get() and get_mut()
+{
+    let mut verts_sorted_by_distance: Vec<NodeIndex> = Vec::new(); // a stack
+    let c = graph.node_count();
+    let mut predecessors =
+        HashMap::<G::NodeId, Vec<G::NodeId>>::with_capacity(c);
+    let mut sigma = HashMap::<usize, i64>::with_capacity(c);
+    let mut distance = HashMap::<G::NodeId, i64>::with_capacity(c);
+    #[allow(non_snake_case)]
+    let mut Q: VecDeque<NodeIndex> = VecDeque::with_capacity(c);
+
+    let i_s = graph.to_index(*node_s);
+    let index_s = NodeIndex::new(i_s);
+
+    for node in graph.node_identifiers() {
+        let i = graph.to_index(node);
+        let index = NodeIndex::new(i);
+        predecessors.insert(index, Vec::new());
+        sigma.insert(i, 0);
+        distance.insert(index, -1);
```

Comment:
there is no need to create a new `index` since `node` is the same.
```suggestion
        let i = graph.to_index(node);
        predecessors.insert(node, Vec::new());
        sigma.insert(i, 0);
        distance.insert(node, -1);
```